feat: allow entering non-default values in multi-select

joobisb opened the pull request:
Closes https://github.com/coder/coder/issues/15488

cliui.Multiselect currently does not allow users to input custom options if zero choices are provided.


### Proposed Solution

cliui.MultiSelect should always allow users to specify another option.

For example, when running coder exp prompt-example multi-select: 

PS: I have modified the text a bit from the original issue

```
? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
> [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  ... Other
+ Add custom value


? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
? Enter custom value: potato⏎


? Select some things:  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
  [x] Code
  [ ] Chair
  [ ] Whale
  [ ] Diamond
  [ ] Carrot
  [x] potato
>  ... Other
+ Add custom value

```

Attaching a recording for the same

https://github.com/user-attachments/assets/ce02cfab-4549-4431-b0d7-c7c2d69ae7ca

joobisb committed:
feat: allow entering non-default values in multi-select

joobisb commented on cli/cliui/select_test.go:
```diff
@@ -101,6 +101,39 @@ func TestMultiSelect(t *testing.T) {
 		require.Equal(t, items, <-msgChan)
 	})
+
+	t.Run("MultiSelectWithCustomInput", func(t *testing.T) {
```
couldn't add tests to check the custom input flow, we had this check here and not sure if its safe to remove and add interactive tests, please do let me know if there's a better way to incorporate those tests

https://github.com/coder/coder/blob/63572d9f5370a860653adf9ff61256a7aed1724c/cli/cliui/select.go#L310

johnstcn replied:
I think historically we've manually tested this using `prompt-example`. I agree that it's not ideal :-(

joobisb commented:
Hi @johnstcn ,

could you please have a look at this PR?

I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?

johnstcn commented:
> I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?

I think it would make more sense to either de-duplicate or validate that the custom option isn't already present.

johnstcn commented on cli/cliui/select.go:
```diff
@@ -370,12 +372,15 @@ type multiSelectOption struct {
+	message           string
+	canceled          bool
+	selected          bool
+	isInputMode       bool   // New field to track if we're adding a custom option
```
suggestion: `isCustomInputMode` for consistency

DanielleMaywood commented on cli/cliui/select.go:
```diff
@@ -414,15 +428,17 @@ func (m multiSelectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 
 		case tea.KeyUp:
 			options := m.filteredOptions()
+			maxIndex := len(options)
 			if m.cursor > 0 {
 				m.cursor--
 			} else {
-				m.cursor = len(options) - 1
+				m.cursor = maxIndex
```
I think this logic needs to handle when `EnableCustomInput` is false. Currently it lets you go over the last valid option.

johnstcn commented on cli/prompts.go:
```diff
@@ -156,9 +156,10 @@ func (RootCmd) promptExample() *serpent.Command {
 						},
-						Defaults: []string{"Code"},
+						Defaults:          []string{"Code"},
+						EnableCustomInput: true,
```
Can we expose this as a flag so we can easily test both with and without custom input?

joobisb committed:
fix: duplicates, cursor positioning

joobisb commented:
> > I do have a question regarding duplicates, if users enters the same custom value twice should we treat it as a single option ? right now it will create 2 duplicate options, what do you think will be the better way to handle that scenario?
> 
> I think it would make more sense to either de-duplicate or validate that the custom option isn't already present.

@johnstcn @DanielleMaywood 

I've addressed the comments and handled duplicates as shown in the recording

- case insensitive comparison - if it exists, it selects and cursor points to the already existing option
- if not add a new option

https://github.com/user-attachments/assets/28d75728-b392-48cc-9a1d-2048c2ab3363

joobisb committed:
chore: remove unwanted change

johnstcn commented on cli/cliui/select.go:
```diff
@@ -457,6 +471,91 @@ func (m multiSelectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+
+	// Check for duplicates
+	for i, opt := range m.options {
+		if strings.EqualFold(opt.option, m.customInput) {
```
I'm not 100% sure about the case-insensitive matching. On the surface, it seems innocuous enough but I'm not sure if there's going to be a case where someone is going to want to differentiate `A` from `a`. Let's keep matching case-sensitive for now.
We can add an option for case-insensitive matching later if someone requests it.

joobisb replied:
updated the PR!

joobisb committed:
remove case insensitive matching
